components/OAuth2.0/authorization.c

```diff
@@ -0,0 +1,283 @@
+    if (!IsResponseReady)
+    {
+        ESP_LOGW(TAG, "timeout - Service not responded!");
+        return false;
+    }
+
+    bool IsTokenExtracted = 
+            ExtractTokenJson(OAuthInterfaceHandler.OAuthBuffer.MessageBuffer,
+                                                    &OAuthInterfaceHandler.token);
+    if (!IsTokenExtracted)
+    {
+        ESP_LOGW(TAG, "new Token not found!");
+        return false;
+    }
+
+    ESP_LOGI(TAG, "new Token found!");
+    OAuthInterfaceHandler.TokenLastUpdate = xTaskGetTickCount();
+    return true;
+}
+
+static void Oauth_MainTask(void *pvparameters)
+{
+    while (1)
+    {
+        char receivedData[MEDIUM_BUF];
+        switch (OAuthInterfaceHandler.Status)
+        {
+            case INIT:
+            {
+                bool IsServerInit = HttpServerServiceInit();                                                     // Init the webserver in initialization
+                if (!IsServerInit)
+                {
+                    ESP_LOGE(TAG, "Authorization initialization failed!");
+                    break;
+                }
+
+                bool IsSpiffExists = SpiffsExistenceCheck(pInterfaceHandler->ConfigAddressInSpiffs);                      // Check if the refresh token exists in the spiffs
+                if (!IsSpiffExists)
+                {
+                    ESP_LOGE(TAG, "Autorization is needed");
+                    OAuthInterfaceHandler.Status = LOGIN;                                                              // If the refresh token does not exist, set the status to LOGIN 
+                    break;
+                }
+
+                ESP_LOGI(TAG, "RefreshToken found");
+                OAuthInterfaceHandler.Status = EXPIRED;                                                            // If the refresh token exists, set the status to EXPIRED
+                break;
+            }
+            case LOGIN:
+            {
+                bool IsCodeReceived = xQueueReceive(SendCodeFromHttpToTask, receivedData, pdMS_TO_TICKS(SEC));       // Waiting for Code to be recieved by queue
+                if (!IsCodeReceived)
+                {
+                    // stay in this state until receive the Code
+                    break;
+                }
+                SendTokenRequest(receivedData);                                                         // send request for Token
+                OAuthInterfaceHandler.Status = AUTHENTICATED;                                                          // Code received and checked, so update Status to AUTHENTICATED         
+                break;
+            }
+            case AUTHENTICATED:
+            {
+                ESP_LOGI(TAG, "AUTHENTICATED");
+                bool IsServiceResponded = xSemaphoreTake(OAuthInterfaceHandler.OAuthBuffer.ResponseReadyFlag, pdMS_TO_TICKS(SEC));              // Waiting for Token to be recieved by queue
+                if (!IsServiceResponded)
```

Comment:
you don't create semaphore, this is the reason that your program doesn't work.

Reply:
It was created in taskinit func.